Add a "pressed" style to PopupMenu

Calinou opened the pull request:
This makes PopupMenus (and derivatives such as OptionButtons) feel more responsive to user input, as they can display a different stylebox and font color when an option is pressed.

**Testing project:** [test_popupmenu_pressed.zip](https://github.com/godotengine/godot/files/11253559/test_popupmenu_pressed.zip)

## Preview

*Note: This `pressed` style only applies to mouse input, not keyboard input (as keyboard inputs are accepted as soon as you press the key, not when you release it). In the videos, I use keyboard input near the end to show the difference.*

### Editor - Dark theme

https://user-images.githubusercontent.com/180032/232256581-5e224e96-ba26-4676-bb4e-3236e1e19cb2.mp4

### Editor - Light theme

https://user-images.githubusercontent.com/180032/232256582-2b4c18ab-7e0b-432d-ab93-26c9c24ea866.mp4

### Default project theme

https://user-images.githubusercontent.com/180032/232256580-a7cee8e7-4b75-451a-afc6-1e3eb7f64bbf.mp4

<details>
<summary>July 2020 patch</summary>

```patch
From d2cf3e17a83ca8d75965c54582f14acaa1ace578 Mon Sep 17 00:00:00 2001
From: Hugo Locurcio <hugo.locurcio@hugo.pro>
Date: Wed, 29 May 2019 05:14:37 +0200
Subject: [PATCH] Add a "pressed" style to PopupMenu

This makes PopupMenus (and derivatives such as OptionButtons) feel
more responsive to user input, as they can display a different
stylebox and font color when an option is pressed.
---
 editor/editor_themes.cpp                      |  9 ++++--
 scene/gui/popup_menu.cpp                      | 29 +++++++++++++++----
 scene/gui/popup_menu.h                        |  1 +
 .../resources/default_theme/default_theme.cpp |  7 +++++
 4 files changed, 39 insertions(+), 7 deletions(-)

diff --git a/editor/editor_themes.cpp b/editor/editor_themes.cpp
index a93763810b5e..f44de9b1b802 100644
--- a/editor/editor_themes.cpp
+++ b/editor/editor_themes.cpp
@@ -556,7 +556,11 @@ Ref<Theme> create_editor_theme(const Ref<Theme> p_theme) {
 
 	Ref<StyleBoxFlat> style_menu_hover_bg = style_widget->duplicate();
 	style_menu_hover_bg->set_border_width_all(0);
-	style_menu_hover_bg->set_bg_color(dark_color_1);
+	style_menu_hover_bg->set_bg_color(dark_color_1.lightened(dark_theme ? 0.01 : 0.0));
+
+	Ref<StyleBoxFlat> style_menu_pressed_bg = style_widget->duplicate();
+	style_menu_pressed_bg->set_border_width_all(0);
+	style_menu_pressed_bg->set_bg_color(dark_color_2.darkened(0.06));
 
 	theme->set_stylebox("normal", "MenuButton", style_menu);
 	theme->set_stylebox("hover", "MenuButton", style_menu);
@@ -566,7 +570,7 @@ Ref<Theme> create_editor_theme(const Ref<Theme> p_theme) {
 
 	theme->set_stylebox("normal", "PopupMenu", style_menu);
 	theme->set_stylebox("hover", "PopupMenu", style_menu_hover_bg);
-	theme->set_stylebox("pressed", "PopupMenu", style_menu);
+	theme->set_stylebox("pressed", "PopupMenu", style_menu_pressed_bg);
 	theme->set_stylebox("focus", "PopupMenu", style_menu);
 	theme->set_stylebox("disabled", "PopupMenu", style_menu);
 
@@ -662,6 +666,7 @@ Ref<Theme> create_editor_theme(const Ref<Theme> p_theme) {
 
 	theme->set_color("font_color", "PopupMenu", font_color);
 	theme->set_color("font_color_hover", "PopupMenu", font_color_hl);
+	theme->set_color("font_color_pressed", "PopupMenu", font_color_hl);
 	theme->set_color("font_color_accel", "PopupMenu", font_color_disabled);
 	theme->set_color("font_color_disabled", "PopupMenu", font_color_disabled);
 	theme->set_icon("checked", "PopupMenu", theme->get_icon("GuiChecked", "EditorIcons"));
diff --git a/scene/gui/popup_menu.cpp b/scene/gui/popup_menu.cpp
index 6e19b820e0c8..8145702fbc9f 100644
--- a/scene/gui/popup_menu.cpp
+++ b/scene/gui/popup_menu.cpp
@@ -285,11 +285,20 @@ void PopupMenu::_gui_input(const Ref<InputEvent> &p_event) {
 	Ref<InputEventMouseButton> b = p_event;
 
 	if (b.is_valid()) {
+		const int button_idx = b->get_button_index();
+
 		if (b->is_pressed()) {
+			if (button_idx == BUTTON_LEFT) {
+				is_mouse_down = true;
+				// Update to display the "pressed" stylebox
+				control->update();
+			}
+
 			return;
+		} else {
+			is_mouse_down = false;
 		}
 
-		int button_idx = b->get_button_index();
 		switch (button_idx) {
 			case BUTTON_WHEEL_DOWN: {
 				_scroll(-b->get_factor(), b->get_position());
@@ -329,8 +338,6 @@ void PopupMenu::_gui_input(const Ref<InputEvent> &p_event) {
 				}
 			}
 		}
-
-		//control->update();
 	}
 
 	Ref<InputEventMouseMotion> m = p_event;
@@ -421,6 +428,7 @@ void PopupMenu::_draw() {
 
 	Ref<StyleBox> style = get_theme_stylebox("panel");
 	Ref<StyleBox> hover = get_theme_stylebox("hover");
+	Ref<StyleBox> pressed = get_theme_stylebox("pressed");
 	Ref<Font> font = get_theme_font("font");
 	// In Item::checkable_type enum order (less the non-checkable member)
 	Ref<Texture2D> check[] = { get_theme_icon("checked"), get_theme_icon("radio_checked") };
@@ -438,6 +446,7 @@ void PopupMenu::_draw() {
 	Color font_color_disabled = get_theme_color("font_color_disabled");
 	Color font_color_accel = get_theme_color("font_color_accel");
 	Color font_color_hover = get_theme_color("font_color_hover");
+	Color font_color_pressed = get_theme_color("font_color_pressed");
 	float font_h = font->get_height();
 
 	// Add the check and the wider icon to the offset of all items.
@@ -477,7 +486,8 @@ void PopupMenu::_draw() {
 		}
 
 		if (i == mouse_over) {
-			hover->draw(ci, Rect2(item_ofs + Point2(-hseparation, -vseparation / 2), Size2(get_size().width - style->get_minimum_size().width + hseparation * 2, h + vseparation)));
+			Ref<StyleBox> hover_pressed = is_mouse_down ? pressed : hover;
+			hover_pressed->draw(ci, Rect2(item_ofs + Point2(-hseparation, -vseparation / 2), Size2(get_size().width - style->get_minimum_size().width + hseparation * 2, h + vseparation)));
 		}
 
 		String text = items[i].xl_text;
@@ -524,7 +534,16 @@ void PopupMenu::_draw() {
 			}
 		} else {
 			item_ofs.x += icon_ofs + check_ofs;
-			font->draw(ci, item_ofs + Point2(0, Math::floor((h - font_h) / 2.0)), text, items[i].disabled ? font_color_disabled : (i == mouse_over ? font_color_hover : font_color));
+
+			Color item_font_color;
+			if (is_mouse_down && i == mouse_over) {
+				item_font_color = font_color_pressed;
+			} else if (i == mouse_over) {
+				item_font_color = font_color_hover;
+			} else {
+				item_font_color = font_color;
+			}
+			font->draw(ci, item_ofs + Point2(0, Math::floor((h - font_h) / 2.0)), text, items[i].disabled ? font_color_disabled : item_font_color);
 		}
 
 		if (items[i].accel || (items[i].shortcut.is_valid() && items[i].shortcut->is_valid())) {
diff --git a/scene/gui/popup_menu.h b/scene/gui/popup_menu.h
index d03a14d6e4e3..ad6ae8d4e52b 100644
--- a/scene/gui/popup_menu.h
+++ b/scene/gui/popup_menu.h
@@ -82,6 +82,7 @@ class PopupMenu : public Popup {
 	Vector<Item> items;
 	int initial_button_mask;
 	bool during_grabbed_click;
+	bool is_mouse_down = false;
 	int mouse_over;
 	int submenu_over;
 	Rect2 parent_rect;
diff --git a/scene/resources/default_theme/default_theme.cpp b/scene/resources/default_theme/default_theme.cpp
index 83d4db7bae69..6fdad3833892 100644
--- a/scene/resources/default_theme/default_theme.cpp
+++ b/scene/resources/default_theme/default_theme.cpp
@@ -509,6 +509,11 @@ void fill_default_theme(Ref<Theme> &theme, const Ref<Font> &default_font, const
 		selected->set_expand_margin_size(Margin(i), 2 * scale);
 	}
 
+	Ref<StyleBoxTexture> pressed = make_stylebox(button_pressed_png, 6, 6, 6, 6);
+	for (int i = 0; i < 4; i++) {
+		pressed->set_expand_margin_size(Margin(i), 2 * scale);
+	}
+
 	theme->set_stylebox("panel", "PopupPanel", style_pp);
 
 	// PopupDialog
@@ -521,6 +526,7 @@ void fill_default_theme(Ref<Theme> &theme, const Ref<Font> &default_font, const
 	theme->set_stylebox("panel", "PopupMenu", style_pd);
 	theme->set_stylebox("panel_disabled", "PopupMenu", make_stylebox(popup_bg_disabled_png, 4, 4, 4, 4));
 	theme->set_stylebox("hover", "PopupMenu", selected);
+	theme->set_stylebox("pressed", "PopupMenu", pressed);
 	theme->set_stylebox("separator", "PopupMenu", make_stylebox(vseparator_png, 3, 3, 3, 3));
 	theme->set_stylebox("labeled_separator_left", "PopupMenu", make_stylebox(vseparator_png, 0, 0, 0, 0));
 	theme->set_stylebox("labeled_separator_right", "PopupMenu", make_stylebox(vseparator_png, 0, 0, 0, 0));
@@ -537,6 +543,7 @@ void fill_default_theme(Ref<Theme> &theme, const Ref<Font> &default_font, const
 	theme->set_color("font_color_accel", "PopupMenu", Color(0.7, 0.7, 0.7, 0.8));
 	theme->set_color("font_color_disabled", "PopupMenu", Color(0.4, 0.4, 0.4, 0.8));
 	theme->set_color("font_color_hover", "PopupMenu", control_font_color);
+	theme->set_color("font_color_pressed", "PopupMenu", control_font_color);
 
 	theme->set_constant("hseparation", "PopupMenu", 4 * scale);
 	theme->set_constant("vseparation", "PopupMenu", 4 * scale);

```
</details>

YeldhamDev commented:
Never seem menus do that in other pieces of software, but sure, why not.

akien-mga commented:
I'm not against the feature and the implementation is fine, but I assume that this is a breaking change for existing games using PopupMenu, no? If they used custom styles for the "panel" and "hover" styleboxes, will this automatically show the default "pressed" stylebox, thus breaking their UI?

Calinou commented:
@akien-mga It will, therefore, it's a breaking change (that has only visual consequences, but still).

girng commented:
i've seen in some popup menus, but rarely.  if it could get added as a non breaking change, it'd be cool i guess. could add to the responsiveness and user experience. just my 2 cents

akien-mga commented:
I think this should likely be optional so that it's not a breaking change. While it can be convenient, it's still a relatively uncommon feature in UIs, so I wouldn't want to force all users to duplicate their stylebox from "normal" to "pressed" when themeing their UI.

Calinou commented:
@akien-mga That makes sense, but I'm not sure how this could be handled in a backwards-compatible manner. (I'm not sure if we've already done something like this before, as we don't add new styles to Theme often.)

akien-mga commented:
> That makes sense, but I'm not sure how this could be handled in a backwards-compatible manner. (I'm not sure if we've already done something like this before, as we don't add new styles to Theme often.)

I would say it could be a boolean option, off by default (for compat) but enabled in the editor theme.
If it's enabled, the `pressed` style property shows up in the Inspector and will be used in the theme, otherwise it uses `normal`.

mrcdk commented:
Wouldn't `if(pressed.is_valid()) {// do pressed and hover} else { // do hover }` work?

akien-mga commented:
@mrcdk That would work too. As long as it's properly documented, that's nicer than a boolean option.

Calinou commented:
I managed to add a backwards compatibility check using `pressed.is_valid()`, but I'm not sure how to do the same for the pressed text color. (When the color doesn't exist in the theme, black is automatically used, which is unreadable.)

aaronfranke commented:
@Calinou Is this still desired? If so, it needs to be rebased on the latest master branch.

Calinou commented:
Rebased. We probably don't need to bother about adding a fallback since we can merge this in the `master` branch without cherry-picking it.

aaronfranke commented:
Marking as Draft because we should likely merge #51159 first and rebase this PR on top of that.

Anutrix commented:
We can move this out of draft and resolve this PR since #51159 is merged.

Calinou commented:
> We can move this out of draft and resolve this PR since #51159 is merged.

Sounds good. I'll probably need to redo this PR from scratch as PopupMenu was refactored in 4.0, so it may take some time.

Calinou committed:
Add a `pressed` style to PopupMenu

This makes PopupMenus (and derivatives such as OptionButtons) feel
more responsive to user input, as they can display a different
stylebox and font color when an option is pressed.

Calinou commented:
Redone for 4.0, with more robust handling when you leave the mouse while pressing the mouse and going back into a PopupMenu and OptionButton (see the default project theme video).

KoBeWi commented:
Tested this and seems like it doesn't work?
![godot windows editor dev x86_64_3cP1BZ6tJT](https://user-images.githubusercontent.com/2223172/232349319-8dc5b629-150c-40ad-bb6d-24eecae9303d.gif)

Calinou commented:
Do you have single-window mode enabled? I've only tested this PR in the editor with single-window mode disabled, or within ([test_popupmenu_pressed.zip](https://github.com/godotengine/godot/files/11253559/test_popupmenu_pressed.zip) (which has Embed Subwindows enabled).

KoBeWi commented:
No, it's non-embedded mode. Not sure why would it matter tho.

Mickeon commented:
This probably sounded good when it was first made. For me personally, almost 5 years later, I don't personally see much of a benefit for this, unless it's an exceedingly requested feature.

Out of curiosity I've opened similar menus, "right-clicked" inside a bunch of applications and **none** of them have a special style for when any of their options is pressed.

Side note, the Godot Editor itself lacks a lot of "feedback" from doing simple tasks such as selecting these options. Perhaps that's the bigger issue here.

fire commented:
Thanks for your work. Let's close and revisit if the mood of the proposal changes.

@jordo, I remember you had some opinions about this; we talked in conversation.

jordo commented:
> Thanks for your work. Let's close and revisit if the mood of the proposal changes.
> 
> @jordo, I remember you had some opinions about this; we talked in conversation.

Perhaps yes, but honestly at this time I don't recall having a strong opinion one way or the other.

KoBeWi commented:
I agree with the above comments. Popup menus don't use pressed style (I'd like to see at least one example that does this :P).

Feel free to open a proposal if you are still interested in this feature, but right now I'm closing it. Thanks for the contribution nonetheless.
